### PR TITLE
[4.13.13f] Adding TLS and eaccess via port 7910

### LINKED ISSUE #73: Simutronics onsite team connection questions.
Lyneya (Nora) has asked me to post this here and if anyone has any questions to forward to the onsite team, they can be sent to her through discord.  Her discord handle is Lyneya#1483.

```
The service that we want to discontinue listens on port 7900 at eaccess.play.net.  We probably have multiple domain names that point to that IP address.  Information exchanged to that service is not protected to any modern standard of encryption.

The replacement service listens on the same IP address but port 7910.  It uses the TLS protocol standard.  The data sent over the encrypted channel is the same format as before except there are no newlines at the end of messages (we rely on TLS to deliver complete messages).

It is important that clients verify that the certificate received by the handshake is what we've configured on the server (to prevent man in the middle attacks).  We use a self-signed certificate, so you can't follow the certificate chain to a trusted root certificate.  In our Windows client, we include the certificate in a file that is read and compared with the received certificate.

That is the high level technical description of what needs to be done.  Beyond that, we can try to answer questions and provide sample C++ code from what we do in the SGE.
```

### PR DESCRIPTION
**Pull request by MahtraDR**
Resolves https://github.com/dragon-realms/dr-lich/issues/73

Major credit to @OSXLich-Doug who babysat me through this, and who really did most of the work...

I've tested on linux, and on windows, I can log in and play. On Windows it launches Wrayth just as expected.

Need more testers... 

Doesn't truly depend on previous PRs, but is built on top of https://github.com/dragon-realms/dr-lich/pull/72

### CODE REVIEW

**Comment by KatoakDR** [comment]
@MahtraDR thank you for taking this initiative!

File: lich.rbw [hunk 1]
```diff
@@ -10757,6 +10758,143 @@ if defined?(Gtk)
...
 end
 
+module EAccess
+  PEM = File.join("#{DATA_DIR}/", "simu.pem")
+#  pp PEM
```
Comment (KatoakDR) [inline lich.rbw:10763]: Not sure what `pp PEM` is for? Can we remove this line?

Comment (MahtraDR) [reply lich.rbw]: Ah, it's pretty print debugging, yes. https://stackoverflow.com/questions/2112016/formatting-rubys-prettyprint

Comment (KatoakDR) [reply lich.rbw]: I guess leave them? It's not my style to have commented out code like that, but I'm also thankful we didn't have to invent this code either :)

**Comment by KatoakDR** [comment]
@MahtraDR I'm attaching a copy of the changes in this PR but with the new code's indentation updated to match the surrounding code blocks they were added to. Figured this was easier than me continuing to make lots of comments.

Why I bring it up is because if we fix it now, while we're aware of it, then we or the next person who may be debugging these lines of code don't have to go through the exercise of trying to line up code blocks. This tripped me up when I was reviewing because I thought a section was ending or continuing some other if/else block but it was actually just outdented too far.

[lich.rbw](https://github.com/dragon-realms/dr-lich/files/8145997/lich.txt)

**Comment by MahtraDR** [comment]
> @MahtraDR I'm attaching a copy of the changes in this PR but with the new code's indentation updated to match the surrounding code blocks they were added to. Figured this was easier than me continuing to make lots of comments.
> 
> Why I bring it up is because if we fix it now, while we're aware of it, then we or the next person who may be debugging these lines of code don't have to go through the exercise of trying to line up code blocks. This tripped me up when I was reviewing because I thought a section was ending or continuing some other if/else block but it was actually just outdented too far.
> 
> [lich.rbw](https://github.com/dragon-realms/dr-lich/files/8145997/lich.txt)

It's the current version.

**Comment by KatoakDR** [comment]
Thanks!

**Comment by KatoakDR** [comment]
I connected to DR and Lich 🎉

**Comment by asechrest** [comment]
Connects fine to DR and Lich for me.

**Comment by KatoakDR** [comment]
@MahtraDR Once https://github.com/dragon-realms/dr-lich/pull/72 is merged, would you mind merging `master` branch into this PR to ensure it's current? Thanks

**Comment by MahtraDR** [comment]
> @MahtraDR Once #72 is merged, would you mind merging `master` branch into this PR to ensure it's current? Thanks

Of course, no worries.

**Review by KatoakDR — APPROVED** [review]

**Comment by KatoakDR** [comment]
@rpherbig This is ready for merge when you're ready, thanks

**Comment by MahtraDR** [comment]
webhook test

File: lich.rbw [hunk 2]
```diff
@@ -10760,6 +10756,143 @@ if defined?(Gtk)
...
+  PACKET_SIZE = 8192
+
+  def self.pem_exist?
+    File.exist? PEM
+  end
+
+  def self.download_pem(hostname = "eaccess.play.net", port = 7910)
```
Comment (vtcifer) [inline lich.rbw:10768]: I *really* don't like the idea of fetching the certificate we're going to use to authenticate/validate the connection to the same server.

**Review by vtcifer — CHANGES_REQUESTED** [review]

File: lich.rbw (hunk 2 above)
Comment (MahtraDR) [reply lich.rbw:10768]: Note the way the internet normally handles self-signed certs is to just not validate at all, which is equally valid here, given the choices Simu made.

Comment (rpherbig) [reply lich.rbw:10768]: Aside from gently asking Simu to do it a better way, I don't really have a better idea. Checking is better than not, even if not by much.

Comment (vtcifer) [reply lich.rbw:10768]: We know what it is currently, and it's expiry is...long. I'm more of the mind of hard-coding it in (like done with lnet) for the time being.  

I've requested to the PM of DragonRealms that they should use the already existing CA signed wildcard cert they have for the website in it's place (we would inherit the CRL, the hostname validation, etc if they do that), but until that's done, my personal preference is validating to what we've been given and we know is currently valid.

Comment (vtcifer) [reply lich.rbw:10768]: This (hardcoded cert check) is what is done with SGE currently btw, so presumably we'd have some notice if they change it in the form of an announcement and an SGE update.

Comment (rcuhljr) [reply lich.rbw:10768]: We could also include the file locally with any distributions and if it resorts to the download do the standard user confirmation spiel like when adding a fingerprint from a server you connect to. I can't see the world where they can hijack communication to the server but can't also download the publicly available cert to use so not sure how much effort it's worth worrying over for now.

Comment (MahtraDR) [reply lich.rbw:10768]: I'm leaving it as is. I think it's a reasonable approach given Simu's choices. Whenif Simu changes, we adapt to that.

Comment (OSXLich-Doug) [reply lich.rbw:10768]: Briefly weighing in - don't care for it myself.  Given all the faults discussed, I'd leave it.  Two reasons.  
1) I don't have to rebundle lich if it gets changed, and 
2) we won't know if it gets revoked and reissued short of watching SGE / news - but we also don't care if a new and valid cert is available.

**Pull request merged**